core/network/src/consensus_gossip.rs

```diff
@@ -40,41 +40,57 @@ struct PeerConsensus<H> {
 	is_authority: bool,
 }
 
-#[derive(Clone, Copy)]
-enum Status {
-	Live,
-	Future,
-}
-
 struct MessageEntry<B: BlockT> {
-	message_hash: B::Hash,
 	topic: B::Hash,
 	message: ConsensusMessage,
 	timestamp: Instant,
-	status: Status,
 }
 
 /// Message validation result.
 pub enum ValidationResult<H> {
 	/// Message is valid with this topic.
 	Valid(H),
-	/// Message is future with this topic.
-	Future(H),
 	/// Invalid message.
 	Invalid,
 	/// Obsolete message.
 	Expired,
 }
 
+/// Action to take on status message.
+pub enum StatusResult<H> {
+	/// Do nothing.
+	Ignore,
+	/// Broadcast all messages with given topics.
+	BroadcastTopics(Vec<H>),
+}
+
 /// Validates consensus messages.
 pub trait Validator<H> {
+	/// New peer is connected.
+	fn new_peer(&self, _who: NodeIndex, _roles: Roles) {
+	}
+
+	/// New connection is dropped.
+	fn peer_disconnected(&self, _who: NodeIndex) {
+	}
+
 	/// Validate consensus message.
 	fn validate(&self, data: &[u8]) -> ValidationResult<H>;
 
+	/// Process status message.
+	fn on_status(&self, _who: NodeIndex, _data: &[u8]) -> StatusResult<H> {
+		StatusResult::Ignore
+	}
+
+	/// Filter out departing messages.
+	fn should_send_to(&self, _who: NodeIndex, _topic: &H, _data: &[u8]) -> bool {
```

Comment:
a bit unfortunate that we've got to re-decode all messages each time we try to send them out.

In GRANDPA we'll likely do it by topic.